src/com/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -883,6 +884,106 @@
     (sql/do-commands
      "DROP INDEX fact_values_string_trgm")))
 
+(defn- lift-fact-paths-into-facts
+  "Pairs paths and values directly in facts, i.e. change facts from (id
+  value) to (id path value)."
+  []
+  (sql/do-commands
+   (if (scf-utils/postgres?)
+     "SET CONSTRAINTS ALL DEFERRED" ;; In case it allows additional optimization?
+     "SELECT 1")
+
+   ;; Build complete facts table as of migration 28.
+   "CREATE TABLE facts_transform
+        (factset_id bigint NOT NULL,
+         fact_path_id bigint NOT NULL,
+         fact_value_id bigint NOT NULL)"
+
+   ;; Needed for the insert if nothing else.
+   "CREATE INDEX fact_paths_path_idx on fact_paths(path)"
+   "CREATE INDEX fact_values_value_hash_idx on fact_values(value_hash)"
+
+   ;; Patch up facts refrences to refer to the min id path/value
+   ;; wherever there's more than one option.
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
```

Comment:
@rbrw I'm having difficulty getting this to finish. I loaded up a database using Wyatt's test functions in benchmark.clj. It had about 430k fact paths and fact values, 720k facts. From the params, it was 2000 nodes, 100 facts per node, 50% duplicate. Thus far the migration has been running for about 20 minutes, not yet finished. Querying PG, it looks like it's running this query. Running explain on it I see (cost=61731.98..1371353952.47 rows=728159 width=85).

Am I doing something wrong? It doesn't look like this will finish for me.